lib/wasi/src/state.rs

```diff
@@ -20,75 +20,297 @@ use wasmer_runtime_core::debug;
 /// the number of symlinks that can be traversed when resolving a path
 pub const MAX_SYMLINKS: u32 = 128;
 
-#[derive(Debug)]
-pub enum WasiFile {
-    HostFile(fs::File),
+/// Error type for external users
+#[derive(Debug, PartialEq, Eq)]
+#[allow(dead_code)]
+// dead code beacuse this is for external use
+pub enum WasiFsError {
+    /// The fd given as a base was not a directory so the operation was not possible
+    BaseNotDirectory,
+    /// Expected a file but found not a file
+    NotAFile,
+    /// The fd given was not usable
+    InvalidFd,
+    /// File exists
+    AlreadyExists,
+    /// Something failed when doing IO. These errors can generally not be handled.
+    /// It may work if tried again.
+    IOError,
+    /// A WASI error without an external name.  If you encounter this it means
+    /// that there's probably a bug on our side (maybe as simple as forgetting to wrap
+    /// this error, but perhaps something broke)
+    UnknownError(__wasi_errno_t),
+}
+
+impl WasiFsError {
+    pub fn from_wasi_err(err: __wasi_errno_t) -> WasiFsError {
+        match err {
+            __WASI_EBADF => WasiFsError::InvalidFd,
+            __WASI_EEXIST => WasiFsError::AlreadyExists,
+            __WASI_EIO => WasiFsError::IOError,
+            _ => WasiFsError::UnknownError(err),
+        }
+    }
 }
 
-impl WasiFile {
-    pub fn close(self) {}
+/// This trait relies on your file closing when it goes out of scope via `Drop`
+/// it does not require `Drop` because `std::fs::File` does not implement it
+/// (TODO: figure out how that makes sense, because it clearly does)
```

Comment:
std::fs::File does not implement Drop, because it contains a std::sys::fs::File, which does not implement Drop as it contains a std::sys::fd::FileDesc, which implements Drop.

aka, only the innermost item needs to implement Drop, rust figures out the rest on it's own?

Reply:
Thanks!  I was reading through the code, but didn't get that far -- that's good to know

Reply:
so, yeah, apparently we should (almost?) never require an impl of Drop, because of the above.